qdrant-landing/content/articles/sayt.md

```diff
@@ -0,0 +1,122 @@
+    }))
+}
+```
+
+If the need arises, one could instead encode the names as UUID, which needs to time-consumingly hash the input, so I decided against this for now.
+
+The `recommend` endpoint works roughly the same as `search_points`, but instead of searching for a vector, we search for one or more points (we can also give negative example points the search will try to avoid). It was built to help drive recommendation engines, saving the roundtrip of sending the current point's vector back to Qdrant to find more similar ones. The endpoint however goes a bit further by allowing us to select a different collection to lookup the points, which allows us to keep our `prefix_cache` collection separate from the site data. So in our case, Qdrant first looks up the point from the `prefix_cache`, takes its vector and searches for that in the `site` collection, essentially caching the precomputed embeddings.
+
+Now we have, in the best Rust tradition, a blazingly fast semantic search.
+
+To demo it, I used our [Qdrant documentation website](https://qdrant.tech/documentation)'s page search, replacing our previous Python implementation. So in order to not just spew empty words, here is a benchmark, showing different queries that exercise different code paths.
+
+Since the operations themselves are far faster than the network whose fickle nature would have swamped most measurable differences, I benchmarked both the Python and Rust services locally. Note that with search terms of up to 3 characters, the Python version merely does a text search, not a semantic search, while the Rust version always does a more complex semantic search. I'm measuring both versions on the same AMD Ryzen 9 5900HX with 16GB RAM running Linux. The table shows the average time and error bound in milliseconds. I only measured up to a thousand concurrent requests. None of the services showed any slowdown with more requests in that range. I do not expect our service to become DDOS'd, so I didn't benchmark with more load. Without further ado, here are the results:
+
+![Benchmark Results](/articles_data/sayt/benchmark.png)
+
+| query     | `io`       | `ring`    | `io_uring` |
+|-----------|------------|-----------|------------|
+| Python 🐍 | 4½ ± ½ ms* | 16 ± 4 ms | 16 ± 4 ms  |
+| Rust   🦀 | 1½ ± ½ ms  | 1½ ± ½ ms |  5 ± 1 ms  |
+
+\* full-text search only
+
+The Rust version consistently beats the Python version in terms of speed and offers a semantic search even on few-character queries. If the prefix cache is hit (as in the "ring" measurement), the semantic search can even get more than ten times faster than the Python version. The general speedup is due to both the relatively lower overhead of Rust + Actix Web compared to Python + FastAPI (even if that already performs admirably), as well as using ONNX Runtime instead of SentenceTransformers for the embedding. The prefix cache gives the Rust version a real boost by doing a semantic search without doing any embedding work.
```

Comment:
I would add here that all those milliseconds are important here for 2 reasons:

- search-as-you-type speed is bound to the user's typing speed. If it is doing to be significantly slower than the keystroke, user will notice.
- search-as-you-type generate search query on each letter. So it usually 3-5x more load than usual search queries.